rfcs/async-helpers.md

```diff
@@ -0,0 +1,279 @@
+# test262 RFC: Async Helpers
+
+- Feature Name: `async-helpers`
+- Start Date: 2022-11-16
+- RFC PR: [tc39/test262#3724](https://github.com/tc39/test262/pull/3724)
+
+## Summary
+
+Helpers that improve the contributor's experience writing tests for the behaviour of asynchronous functions.
+
+## Motivation
+
+There are two proposals coming down the line that will add asynchronous built-in functions to JS.
+`Array.fromAsync` is one, and it is already at Stage 3.
+The other is iterator helpers, which is on the agenda to request Stage 3 for the November/December 2022 plenary meeting. Farther in the future, we may have async do-expressions.
+This all suggests that there will be a fair amount of test coverage added in 2023 and beyond that will need to deal with the behaviour of asynchronous functions.
+
+This RFC proposes some new helpers for these tests.
+The goal is to provide facilities for test writers to easily test things that happen asynchronously, and to make it easier to reason about the asynchronous control flow in such tests at a glance.
+
+## Guide-level explanation
+
+Concretely, this RFC proposes a new helper file that can be included in a test262 test via `includes: [asyncHelpers.js]`.
+
+This is intended to benefit:
+- Test writers, by providing facilities for writing asynchronous tests that are easier to deal with and easier to reason about;
+- test262 maintainers, by letting contributors write tests that avoid boilerplate and are easier to review at a glance.
+
+Implementations that consume the test suite should not be affected, since existing test runners should be able to handle the helper file without any changes to the test runner.
+
+As a usage example in this section, as well as in the Prior Art section below, I'll use the `Array.fromAsync()` equivalent of [`test/built-ins/Array/from/mapfn-is-not-callable-typeerror.js`](https://github.com/tc39/test262/blob/main/test/built-ins/Array/from/mapfn-is-not-callable-typeerror.js), which is written approximately as follows (edits made for the purpose of keeping the RFC brief and consistent):
+```js
+assert.throws(TypeError, () => Array.from([], null), "null mapfn");
+assert.throws(TypeError, () => Array.from([], {}), "non-callable object mapfn");
+assert.throws(TypeError, () => Array.from([], "String"), "string mapfn");
+assert.throws(TypeError, () => Array.from([], true), "boolean mapfn");
+assert.throws(TypeError, () => Array.from([], 3.1416), "number mapfn");
+```
+
+An example test using the async helpers, testing the equivalent step in the Array.fromAsync proposal, might look like this:
+```js
+/*---
+...
+flags: [async]
+includes: [asyncHelpers.js]
+---*/
+
+asyncTest(async function () {
+  await assert.throwsAsync(TypeError, Array.fromAsync([], null), "null mapfn");
```

Comment:
```suggestion
  await assert.throwsAsync(TypeError, () => Array.fromAsync([], null), "null mapfn");
```

And likewise with the other ones below?

Reply:
Yep, this was left over from an earlier draft where a Promise was allowed here. Good catch!